drivers/power/ltc3337/ltc3337.c

```diff
@@ -0,0 +1,715 @@
+	//to be portable across differnt systems:
+	//Let's convert our qlsb_m_nA to whole and fractional parts
+	//This is to prevent an overflow during the math.
+	//Worst case is 14910000 nA with a prescaler of 0. If we had a full counter
+	//(0xFFFF), multiplication works out to 0xE3_814C_7DD0, 32-bit overflow.
+	//Using base-2 components is easier programming wise, but using base-10
+	//helps track the units better:
+	//Divisor of 10,000 (10uA per whole value, 1nA per fract Value)
+	//  - 14910000 / 10000 = 1491 * 0xFFFF (full counter) = 0x5D2_FA2D
+	//  - 9,999 (Wost case remainder) * 0xFFFF = 0x270E_D8F1
+	whole_count = (uint32_t)charge_count * (qlsb_m_na / LTC3337_CALC_SCALE);
+	fract_count = (uint32_t)charge_count * (qlsb_m_na % LTC3337_CALC_SCALE);
+
+	ret_value->a_hr = whole_count /
+			  LTC3337_CALC_TO_WHOLE; //10uA per bit,  / 100,000 to get A
+	ret_value->na_hr = (whole_count % LTC3337_CALC_TO_WHOLE) *
+			   LTC3337_CALC_SCALE; //*10,000 nA per bit
+	ret_value->na_hr += fract_count; //Add in the other nA's
+
+	//Add back in the remainder after shifting out the prescaler
+	//Each LSB is 1/2^M nA
+	fract_count = qlsb_m_remainder * charge_count;
+	ret_value->na_hr += fract_count >> prescale;
+
+	//If the nA-hr component overflowed 1A, adjust accordingly
+	while(ret_value->na_hr > LTC3337_NANO_AMP) {
+		ret_value->a_hr++;
+		ret_value->na_hr -= LTC3337_NANO_AMP;
+	}
+}
+
+/**
+ * Converts a charge in A/nA-hr to an equivalent count register
+ * value to be used for setting alarms and clearing the register
+ * @param prescale - Prescale value
+ * @param lsb_na - LSB value in Nano-amps
+ * @param charge_a - Pointer to the charge value in A/nA-hrs
+ * @returns Equivalent register value based on parameters
+ */
+static uint16_t ltc3337_a_to_charge_count(uint8_t prescale, uint32_t lsb_na,
+		struct charge_count_t* charge_a)
+{
+	uint32_t qlsb_m_na;
+	uint32_t qlsb_m_remainder;
+	uint32_t bits_per_a;
+	uint32_t extra_per_a;
+	uint32_t working_count;
+	uint32_t remainder_count;
+	uint32_t fract_count;
+
+	//Per the data sheet qLSB_M = (qLSB / 2*M)
+	//We can get qLSB_M by doing (qLSB >> M);
+	qlsb_m_na = (lsb_na >> prescale);
+
+	//Get the fractional bits we shifted out
+	qlsb_m_remainder = lsb_na & ((1 << prescale) - 1);
+
+	//Calculate the number of bits for each A-hr
+	bits_per_a = LTC3337_NANO_AMP / qlsb_m_na;
+	extra_per_a = LTC3337_NANO_AMP % qlsb_m_na;
+
+	//Exceeds the bounds of the configuration, and could overflow
+	//the math.  Drop out now.
+	if(charge_a->a_hr > (0xFFFF / bits_per_a))
+		return 0xFFFF;
+
+	//Get the baseline count
+	working_count = (charge_a->a_hr * bits_per_a);
+	working_count += (charge_a->na_hr / qlsb_m_na);
+
+	//Accumulate any remainders from the math, and add extra bits
+	//if we overflow qlsb_m_nA
+	remainder_count = (charge_a->a_hr * extra_per_a);
+	remainder_count += (charge_a->na_hr % qlsb_m_na);
+	working_count += (remainder_count / qlsb_m_na);
```

Comment:
Maybe NO_OS_DIV_U64 can be used for this and other similar division operations in the code.

Reply:
This particular block would not benefit as its working with 2 different data sources na_hr and a_hr.  The code [here ](https://github.com/BrentK-ADI/no-OS/blob/070fd70f744cbd73f496f835cb115925c366bc76/drivers/power/ltc3337/ltc3337.c#L631)
could utilize it, but would require a re-write of that function.  Not sure if value add since the existing implementation was completely unit tested against all possible values.